pkg/ecdsa/dkg/params_pool.go

```diff
@@ -0,0 +1,81 @@
+package dkg
+
+import (
+	"fmt"
+	"github.com/binance-chain/tss-lib/ecdsa/keygen"
+	"github.com/ipfs/go-log"
+	"time"
+)
+
+// tssPreParamsPool is a pool holding TSS pre parameters. It autogenerates
+// entries up to the pool size. When an entry is pulled from the pool it
+// will generate a new entry.
+type tssPreParamsPool struct {
+	logger         log.StandardLogger
+	pool           chan *keygen.LocalPreParams
+	newPreParamsFn func() (*keygen.LocalPreParams, error)
+}
+
+// newTssPreParamsPool initializes a new TSS pre-parameters pool.
+func newTssPreParamsPool(
+	logger log.StandardLogger,
+	poolSize int,
+	generationTimeout time.Duration,
+) *tssPreParamsPool {
+	logger.Infof("TSS pre-parameters target pool size is [%v]", poolSize)
+
+	newPreParamsFn := func() (*keygen.LocalPreParams, error) {
+		preParams, err := keygen.GeneratePreParams(generationTimeout)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to generate TSS pre-params: [%v]",
+				err,
+			)
+		}
+
+		return preParams, nil
+	}
+
+	tssPreParamsPool := &tssPreParamsPool{
+		logger:         logger,
+		pool:           make(chan *keygen.LocalPreParams, poolSize),
+		newPreParamsFn: newPreParamsFn,
+	}
+
+	go tssPreParamsPool.pumpPool()
+
+	return tssPreParamsPool
+}
+
+func (t *tssPreParamsPool) pumpPool() {
+	for {
+		t.logger.Info("generating new tss pre parameters")
+
+		start := time.Now()
+
+		params, err := t.newPreParamsFn()
+		if err != nil {
+			t.logger.Warningf(
+				"failed to generate tss pre parameters after [%s]: [%v]",
+				time.Since(start),
+				err,
+			)
+			continue
+		}
+
+		t.logger.Infof(
+			"generated new tss pre parameters, took: [%s], "+
+				"current pool size: [%d]",
+			time.Since(start),
+			len(t.pool)+1,
+		)
+
+		t.pool <- params
```

Comment:
Oh, because this is `t.pool` is a fixed-size channel at `poolSize`, this blocks when `len(t.pool) == poolSize` until `t.get` is called. That consumes a pre-param from the channel, opens up a slot, then we immediately fill the slot, and start calculating the next one until we get blocked again.

The `for {` style loop infinitely repeats, but doesn't continually execute code because the pool runs out of space and it blocks on that goroutine.